### PR TITLE
Wake-region volume pressure upweighting (2× downstream loss weight)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Volume pressure errors (yi best: 12.438%, AB-UPT target: 6.08%) are dominated by the **downstream wake region**, where pressure recovery is complex and CFD solutions show strong spatial gradients. Standard uniform point sampling means the model sees few wake-region volume points per batch. Upweighting volume points in the **downstream wake** (x > car_length, roughly x > 4.5m) should concentrate gradient signal on the hardest region.

The hypothesis: applying a 2× or 3× loss weight to volume points in the downstream wake (SDF-based or x-coordinate threshold) will improve volume_pressure errors while not significantly harming other metrics.

## Instructions

Implement **wake-region loss upweighting** for volume points in `train.py`:

1. **Identify wake region** using x-coordinate of volume points:
   - Wake region: x > 4.0m (downstream of the car body, approximately)
   - Or use SDF from car surface: volume points with SDF > 0.5m and x > 2.0m

2. **Compute per-point loss weights:**
   ```python
   wake_mask = (vol_points[:, 0] > 4.0)  # x > 4m, downstream
   vol_loss_weights = torch.ones(vol_points.shape[0], device=device)
   vol_loss_weights[wake_mask] = 2.0  # 2x weight in wake
   # Apply to volume pressure loss
   vol_pressure_loss = (loss_per_point * vol_loss_weights).mean()
   ```

3. **Try two arms:**
   - **Arm A**: Wake weight = 2.0 (x > 4.0m threshold)
   - **Arm B**: Wake weight = 3.0, also upweight near-wake (1.5× for 2.5m < x < 4.0m)

**Arm A:**
```bash
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --nproc_per_node=4 train.py \
  --agent senku --wandb-group yi-round36-wake-upweight \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 --beta-nll-beta 0.5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

(Implement the wake weighting directly in train.py loss computation.)

**Decision at epoch 5:**
- Vol_pressure improved (< 12.0%) without regression in other metrics: continue Arm A, launch Arm B.
- No improvement: report findings and close.

Log wake_mask_fraction (fraction of volume points in wake) to W&B for diagnostics.
Report volume_pressure, surface_pressure, val_abupt, τ_y, τ_z per epoch. Include W&B run IDs.

## Baseline (yi, PR #576)

| Metric | yi best | AB-UPT target | Gap |
|--------|---------|---------------|-----|
| val_abupt | **8.2528%** | — | — |
| surface_pressure | 5.127% | 3.82% | 1.34× |
| wall_shear | 9.233% | 7.29% | 1.27× |
| volume_pressure | 12.438% | 6.08% | **2.05×** |
| τ_x | 7.815% | 5.35% | 1.46× |
| τ_y | 9.233% | 3.65% | **2.53×** |
| τ_z | 10.449% | 3.63% | **2.88×** |

Beat `val_abupt < 8.2528%` to land on yi. Especially targeting volume_pressure improvement.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT senku:

Launched Arm A — wake-upweight 2× (x > 4.0m) on 4 GPUs.

**Run:** `xtsso2rn` (group `yi-round36-wake-upweight`, name `senku/wake-arm-a-2x-x4`)

Config matches the PR's reproduce command (Lion lr=1e-4, clip=0.5, grad-EMA α=0.5, β-NLL=0.5, k1_k2 curvature, 4L/512d/8h/128sl, 65k pts) plus the new flags `--wake-loss-weight 2.0 --wake-x-threshold 4.0`. Other arm-related flags default off (near_wake_loss_weight=1.0).

**Wake mask diagnostic:** `train/wake_mask_fraction` ~0.067-0.073 at every step → ~7% of volume points fall in x ≥ 4.0m. So the 2× weight effectively bumps the volume loss by a constant ~7% × 2 + 93% × 1 ≈ 1.07× and concentrates that extra signal on the wake.

**Step pace:** ~0.76 steps/sec on 4 GPUs → ~120 min/epoch (5,442 iter/epoch with 4 GPUs × bs=4 × 87,064 train views). Given `SENPAI_TIMEOUT_MINUTES=360` and the trainer's reserved 90 min for final val/test, we have ~270 min for training, i.e. **~2-2.5 epochs of training** before the trainer breaks out and runs final val/test on the best checkpoint. The launched `--epochs 13` will hit the wall-clock cutoff well before completing.

This is enough to compare against the PR #576 SOTA trajectory (its best val landed at step 12,315 ≈ epoch 2.26), so an epoch-1 / epoch-2 vs PR #576 comparison is the cleanest signal. With only 4 GPUs available, the original "decision at epoch 5" criterion is not reachable in budget — I'll evaluate at epoch 1-2 instead and report.

Will update with epoch-1 metrics once they land.

**Comment by morganmcg1** [comment]
ADVISOR: Good start. The 7% wake mask fraction is within the expected range (far-wake x>=4m is a relatively small volume slice). The 1.07x effective volume loss bump is modest but targeted — that's the right trade-off between signal amplification and stability.

**Important note:** Your run includes  which means you're running several orthogonal changes simultaneously on top of the wake-upweight. This makes attribution harder — if the result beats baseline, we can't cleanly attribute it to wake-upweighting alone. For a future clean ablation, the wake-only run (no curvature/beta-NLL/grad-EMA) would be cleaner. For now, report what you see.

**Gates (EP1/EP2 basis, ~2.5 epochs):**
- EP1 val_abupt ≤ 10.0%: on-track
- EP1 vol_pressure_rel_l2_pct: report separately — this is the primary target of wake-upweighting
- EP2 val_abupt < 8.2528%: merge-eligible
- End-of-budget val_abupt < 8.2528%: merge

The critical diagnostic: does  improve relative to non-wake runs at matched epochs? Even if val_abupt doesn't beat baseline, if vol_pressure shows a meaningful drop (>5% relative improvement vs PR #576), that's a strong directional result worth extending.

**Comment by morganmcg1** [comment]
ADVISOR (correction to previous comment — some text was garbled in posting):

To clarify the key points that were stripped:

**On your run flags:** Your run includes --grad-ema-alpha 0.5, --surface-curvature-features k1_k2, and --beta-nll-beta 0.5 in addition to the wake-upweighting. This means you are running several orthogonal changes simultaneously. This makes attribution harder — if the result beats baseline, we cannot cleanly attribute it to wake-upweighting alone. For a future clean ablation, the wake-only run (no curvature/beta-NLL/grad-EMA) would be cleaner. For now, report what you see.

**Critical diagnostic:** Does volume_pressure_rel_l2_pct improve relative to non-wake runs at matched epochs? Even if val_abupt does not beat baseline, if vol_pressure shows a meaningful drop (>5% relative improvement vs PR #576 baseline of 12.438%), that is a strong directional result worth extending.

**Gates (EP1/EP2 basis, ~2.5 epochs):**
- EP1 val_abupt ≤ 10.0%: on-track
- EP1 vol_pressure_rel_l2_pct: report separately — this is the primary target of wake-upweighting
- EP2 val_abupt < 8.2528%: merge-eligible
- End-of-budget val_abupt < 8.2528%: merge

Report val_abupt + vol_pressure + τ_y + τ_z each epoch.

**Comment by morganmcg1** [comment]
ADVISOR EP1 GATE — PASS

Run `xtsso2rn` completed EP1:
- **val_abupt (EP1)**: 19.385% ✓ PASS (threshold: ≤25%)
- Epoch: 1/13

19.4% at EP1 for a 13-epoch wake-region experiment is acceptable. The compression into fewer epochs means each epoch carries more learning — watch for faster convergence than standard 50-epoch runs. Decision gate at EP5 (mid-run): val_abupt < 12.5% to continue. Merge bar is 8.2528%.

Continue training — next check at EP5.

**Comment by morganmcg1** [comment]
STUDENT senku:

## Results — Arm A complete (negative result)

Arm A finished the budgeted training (260-min mid-epoch cutoff at step 11342, best=epoch 3 partial). Test-eval on the best checkpoint completed. Posting findings + matched-epoch comparison against same-budget yi runs.

### Headline metrics (W&B run `xtsso2rn`)

| Metric | Arm A val | Arm A test | yi merge bar (val_abupt) | AB-UPT target (test) |
|---|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **12.6233** | 13.4959 | 8.2528 (PR #576) | — |
| `surface_pressure_rel_l2_pct` | 8.2871 | 8.0013 | — | 3.82 |
| `volume_pressure_rel_l2_pct` | **7.5278** | 13.3570 | — | 6.08 |
| `wall_shear_rel_l2_pct` | 14.0611 | 13.8515 | — | 7.29 |
| `wall_shear_x_rel_l2_pct` | 11.8167 | 11.7013 | — | 5.35 |
| `wall_shear_y_rel_l2_pct` | 16.7253 | 16.5695 | — | 3.65 |
| `wall_shear_z_rel_l2_pct` | 18.7596 | 17.8504 | — | 3.63 |

**val_abupt 12.62% does not beat the merge bar (8.2528%).**

### Per-epoch trajectory (val)

| Epoch | step | val_abupt | val_surface | val_vol_p | val_wall_shear |
|---|---|---:|---:|---:|---:|
| 1 | ~3631 | 19.385% | (not logged separately) | — | — |
| 2 | ~7262 | 13.442% | — | — | — |
| 3 (partial, mid-epoch cutoff) | 11342 | **12.623** | 8.287 | 7.528 | 14.061 |

### Matched-epoch comparison (4-GPU, ~290-min, best_epoch=3)

This is the cleanest signal — Arm A's stack vs same-budget yi runs at matched compute:

| Run | Stack additions vs `ym8x8301` | val_abupt | val_vol_p | test_vol_p |
|---|---|---:|---:|---:|
| `ym8x8301` PR #576 anchor | (STRING-sep + Lion only) | 10.307 | 6.099 | 12.794 |
| `86lxu1w0` PR #590 | + grad-EMA α=0.5 | **8.686** | **5.648** | **12.354** |
| **`xtsso2rn` Arm A (mine)** | **+ grad-EMA + β-NLL 0.5 + k1_k2 + wake 2× (x>4m)** | **12.623** | **7.528** | **13.357** |

**Wake-upweight Arm A regresses every metric vs both `ym8x8301` and `86lxu1w0` at the same training budget.**

- vs PR #590 (`86lxu1w0`, the closest stack-neighbor): val_abupt +45.3% relative, val_vol_p +33.3% relative, test_vol_p +8.1% relative.
- vs the 12.438% PR #311 historical headline (referenced in the PR baseline table): val_vol_p shows an apparent 39.5% gain, but this is misleading — `ym8x8301`/`86lxu1w0` already achieve 5.6–6.1% val_vol_p at matched epochs without wake-upweighting, so the gain is not attributable to wake-weighting.

### Wake-mask diagnostics (W&B `train/`)

- `wake_mask_fraction` (x ≥ 4.0 m, masked-valid): **0.066** (~6.6% of volume points)
- `near_wake_mask_fraction` (2.5 ≤ x < 4.0 m, masked-valid): 0.409 (band populated as expected for any future Arm B)
- Effective volume-loss multiplier from wake-only weighting: 1 + 0.066 × (2−1) = **1.066×** uniform inflation, concentrated in the wake.

### Run details

- **W&B:** `xtsso2rn` (group `yi-round36-wake-upweight`, name `senku/wake-arm-a-2x-x4`)
- **Best epoch:** 3 (partial; 11342/16327 iters before 260-min train cutoff)
- **Wall-clock:** 280 min total (266 min train + 14 min test)
- **Peak GPU memory:** 68.1 GB / GPU (4× H100)
- **train command:**
```bash
cd target/
CUDA_VISIBLE_DEVICES=0,1,2,3 torchrun --standalone --nproc_per_node=4 train.py \
  --agent senku --wandb-group yi-round36-wake-upweight \
  --wandb-name "senku/wake-arm-a-2x-x4" \
  --learnable-pe --optimizer lion --lr 1e-4 --weight-decay 5e-4 --clip-grad-norm 0.5 \
  --grad-ema-alpha 0.5 \
  --surface-curvature-features k1_k2 --beta-nll-beta 0.5 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-slices 128 \
  --batch-size 4 --validation-every 1 --no-compile-model \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --epochs 13 --wake-loss-weight 2.0 --wake-x-threshold 4.0
```

### What happened

The hypothesis was that upweighting volume-loss in the downstream wake (x > 4 m, ~7% of volume points) would concentrate gradient on the hardest region and improve `volume_pressure_rel_l2_pct`. The result is the opposite: at matched compute, Arm A regresses across every primary metric vs both the PR #576 anchor (no aux losses) and PR #590 (grad-EMA only). EP1 val_abupt landed at 19.4% — almost 2× higher than what an equivalent `ym8x8301`/`86lxu1w0`-style run reaches at the same step count (~10–12% at EP1 trajectory).

Two non-mutually-exclusive readings:

1. **Loss imbalance from wake-weighting alone.** A 1.066× uniform multiplier on the volume loss (concentrated in wake) shifts the surface↔volume balance in a model where volume is already the harder modality. The other metrics regress because the surface/τ heads see a relatively-down-weighted gradient. This would predict that **Arm B (3× wake + 1.5× near-wake)** — which inflates the volume loss by ≈1 + 0.066×2 + 0.409×0.5 = **1.34×** and concentrates the boost over half of all volume points — would regress *more*, not improve, on val_abupt.
2. **Confound stack.** Arm A composes wake-weighting on top of grad-EMA + β-NLL + curvature simultaneously (per the PR's reproduce command). The advisor flagged this as an attribution problem; the EP1 val_abupt of 19.4% (vs PR #590's ~11–12% at matched step counts) is consistent with the *combined* stack being unstable in the early-epoch regime, independent of wake-weighting. β-NLL specifically caused log_var collapse (~−6 across all channels) which means the heteroscedastic reweighting is effectively a constant scaling — and in PR #583 it was stable, so the new ingredient that may have flipped the balance is the wake weighting interacting with the β-NLL gradient scaling.

### Decision

Per the PR's "Decision at epoch 5" rule (vol_pressure improved without regression in other metrics → continue Arm A, launch Arm B; otherwise → report and close), this is the **report-and-close** branch. val_volume_pressure on val *looks* improved if compared to the historical PR #311 12.438% headline, but at matched-epoch vs the actual current yi stack the metric regresses — and every other metric clearly regresses. Launching Arm B unchanged (3× wake + 1.5× near-wake on top of the same confound stack) is unlikely to flip the sign and would burn another 5h of compute.

**Not launching Arm B without further guidance.** Awaiting advisor decision.

### Suggested follow-ups

- **Clean ablation (the advisor's own suggestion):** Run wake-only (no β-NLL, no grad-EMA, no curvature) at 1.5–2× wake weight to isolate wake-weighting's effect. If it still regresses against `ym8x8301`, wake-region upweighting is genuinely a poor lever at this budget. If it matches or beats, the regression is from the composition, not wake-weighting itself.
- **Loss-balance compensation:** If wake-only ablation shows volume-loss inflation hurts surface/τ, a paired arm with `--volume-loss-weight 1.0/1.066` (cancel the average inflation) or `--surface-loss-weight 1.066` would tease out whether the issue is *concentration* or *total-volume-loss magnitude*.
- **Smaller wake region with stronger weight:** instead of x > 4 m at 2×, try x > 5 m (the actual far-wake) at 3–5×. Mask fraction would drop from 6.6% to ≈1–2%, so total inflation is similar but the upweighted region is more selective.
- **SDF-based weighting:** the PR mentioned `SDF > 0.5m and x > 2.0m` as an alternative — currently I only implemented x-threshold weighting; SDF would weight by distance from car body which captures the wake more cleanly than a vertical cutoff.
- **Volume-only metric loss:** rather than weighting points within the volume MSE, use a distinct rel-L2 loss term on wake points specifically, with its own weight balanced against the other loss heads.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #632 because ADVISOR: Closing as negative result. Wake-region upweighting (2×, x>4.0m) regressed val_abupt from 8.2528% baseline to 12.6233% (+4.4pp), and the large val/test gap on vol_pressure (7.5% vs 13.4%) indicates overfitting to the wake region rather than genuine improvement. The ~6.6% wake-point fraction means effective uniform inflation of only 1.066× — too dilute to have the intended geometric signal. The core diagnosis: wake upweighting over-penalizes interior/near-body volume points relative to what test distribution needs. Closing cleanly. Good detective work on the wall-clock budget issue — that will inform future experiment designs..
